apps/settings/js/modules/mvvm/observable.js

```diff
@@ -1,102 +1,353 @@
+    if (typeof observer !== 'function') {
+      return;
+    }
+    (this._observers[name] = this._observers[name] || []).push(observer);
+  };
+
+  /**
+   * Unobserve a property
+   *
+   * @access public
+   * @memberOf Observable.prototype
+   * @param {String} name
+   * @param {Function} observer
+   */
+  Observable.prototype.unobserve = function o_unobserve(name, observer) {
+    if (typeof name === 'function') {
+      // (observer) -- remove from every key in _observers
+      Object.keys(this._observers).forEach(
+        this._removeObserver.bind(this, name));
+    } else {
+      if (observer) {
+        // (prop, observer) -- remove observer from the specific prop
+        this._removeObserver(observer, name);
+      } else if (name in this._observers) {
+        // (prop) -- otherwise remove all observers for property
+        this._observers[name] = [];
+      }
+    }
+  };
+
+  // Static functions
+  var _dependencyGraphs = new Map();
+  /**
+   * Each module should have its own dependency graph that decides what
+   * observers to called when a property changes. The function returns the
+   * dependency graph of a module. It creates one if the map does not exist.
+   *
+   * @param {Object} modulePrototype
+   */
+  function _getDependencyGraph(modulePrototype) {
+    var dependencyGraph = _dependencyGraphs.get(modulePrototype);
+    if (!dependencyGraph) {
+      // register a new dependency graph of the module based on the existing
+      // dependency graph on the module.
+      dependencyGraph = DependencyGraph(modulePrototype._dependencyGraph);
+      modulePrototype._dependencyGraph = dependencyGraph;
+      _dependencyGraphs.set(modulePrototype, dependencyGraph);
+    }
+    return dependencyGraph;
+  }
 
-    var _getFunctionTemplate = function(p) {
-      return function() {
-        return _observable['_' + p];
-      };
+  function _getterTemplate(name, defaultValue) {
+    return function() {
+      var value = this[OP_PREFIX(name)];
+      if (typeof value === 'undefined') {
+        value = this[OP_PREFIX(name)] = defaultValue;
+      }
+      return value;
     };
+  }
+
+  function _setterTemplate(name) {
+    return function(value) {
+      var oldValue = this[name];
+      if (oldValue !== value) {
+        var dependentList = this._dependencyGraph &&
+          this._dependencyGraph.getAllDependent(name);
+        var oldValues;
 
-    var _setFunctionTemplate = function(p) {
-      return function(value) {
-        var oldValue = _observable['_' + p];
-        if (oldValue !== value) {
-          _observable['_' + p] = value;
-          var handlers = _eventHandlers[p];
-          handlers.forEach(function(handler) {
-            handler(value, oldValue);
+        // caches the old values of all dependent
+        if (dependentList && dependentList.length) {
+          oldValues = {};
+          dependentList.forEach((name) => {
+            oldValues[name] = this[name];
```

Comment:
I think this block is quite isolated enough that can be extracted as another function.